SPEC/FILESv2.md

```diff
@@ -0,0 +1,237 @@
+
+### 4. Hoist all methods in the `files` namespace to the root level
+
+Methods that are integral for interacting with the InterPlanetary File System will reside on the root namespace. The reasoning is that these commands are important and will be used often so need to be given priority and ease of access over other APIs that IPFS exposes. It will more effectively advertise what the IPFS core functionality is to aid onboarding and understanding of IPFS in general.
+
+For clarity, the API movement/renaming changes are as follows:
+
+| Old name | New name |
+|---|---|
+| `ipfs add` | `ipfs import` |
+| `ipfs cat` | (removed) |
+| `ipfs get` | `ipfs export` |
+| `ipfs ls` | (removed) |
+| `ipfs files cp` | `ipfs cp` |
+| `ipfs files flush` | `ipfs flush` |
+| `ipfs files ls` | `ipfs ls` |
+| `ipfs files mkdir` | `ipfs mkdir` |
+| `ipfs files mv` | `ipfs mv` |
+| `ipfs files read` | `ipfs read` |
+| `ipfs files rm` | `ipfs rm` |
+| `ipfs files stat` | `ipfs stat` |
+| `ipfs files write` | `ipfs write` |
+
+### 5. Allow both IPFS and MFS paths in API methods
+
+Rather than segregating MFS from the rest of IPFS, we can use MFS paths to refer to content on our local node and IPFS paths to refer to content on the wider IPFS network. We can draw an analogy here with way we use Unix paths and URLs today for working with files in our OS and the Internet respectively. This is natural for many computer users and it's widely understood that the path at the end of a domain navigates through a set of files in much the same way as a path navigates through a set of files on an OS.
+
+Where it makes sense, we can allow MFS paths in the root level API methods and IPFS paths in the MFS API methods. This has already been proven possible as many MFS API methods already accept IPFS paths.
+
+| Method | Accepts IPFS paths | Accepts MFS paths |
+|---|---|---|
+| `ipfs cp` | ✅ | ✅ |
+| `ipfs export` | ✅ | ✅ |
+| `ipfs flush` | ❌ | ✅ |
+| `ipfs import` | ❌ | ❌ |
+| `ipfs ls` | ✅ | ✅ |
+| `ipfs mkdir` | ❌ | ✅ |
+| `ipfs mv` | ✅ | ✅ |
+| `ipfs read` | ✅ | ✅ |
+| `ipfs rm` | ❌ | ✅ |
+| `ipfs stat` | ✅ | ✅ |
+| `ipfs write` | ❌ | ✅ |
+
+The `/ipfs` directory in MFS problem can simply be avoided by either assuming IPFS path (the current solution) or by denying writes to a directory of this name.
+
+### 6. Streaming APIs by default
+
+The file system APIs will be streaming by default. Due to the way we store and retrieve data it makes sense for our API methods to stream content when retrieving it locally or over the network. Buffering APIs can cause OOM issues, give no feedback to the user on progress and they can be trivially wrapped to collect all items in order to achieve the same effect as a buffering API.
+
+Streaming APIs will use a language native / standard library feature that is supported in all runtimes that IPFS is actively targeting. This prevents bloat and by only supporting one streaming mechanism it reduces API surface area.
```

Comment:
Note: Let's also stop streaming _bytes_. If we switch to only streaming _objects_, we can stream metadata (progress, etc.).